File: asg3/hw3.hs
{-
 - Authors: Yunyi Ding and Brian Lin
 -}

data BST k v = Empty | Node k v (BST k v) (BST k v)
val :: BST k v -> Maybe v
val Empty = Nothing
val (Node k v _ _) = Just v

size :: BST k v -> Int
size Empty = 0;
size (Node k v l r) = 1 + size l + size r

File: asg3/hw3.pl
/** 
* Authors: Yunyi Ding and Brian Lin
*/

father(al, bud).
father(al, kelly).
mother(peggy, kelly).
mother(peggy, bud).
mother(martha, peggy).

grandma(X, Y) :- mother(Z, Y), mother(X, Z).

descendants(X, Y) :- mother(X, Y); father(X, Y); grandma(X, Y).

siblings(X, Y) :- mother(M, X), mother(M, Y), father(F, X), father(F, Y), X \= Y.

